controllers/clustertopology_controller_compute.go

```diff
@@ -140,6 +141,149 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
 	return class, nil
 }
 
+// getCurrentState gets information about a cluster topology by inspecting the state of the cluster infrastructure,
+// the control plane, and the machine deployments associated with the cluster.
+func (r *ClusterTopologyReconciler) getCurrentState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (*clusterTopologyState, error) {
+	var err error
+	var infra *unstructured.Unstructured
+	var cp controlPlaneTopologyState
+	var m []machineDeploymentTopologyState
+	// Reference to the cluster infrastructure can be nil and is expected to be on the first reconcile loop
+	// This reference can be nil and the method should still be allowed to continue
+	if cluster.Spec.InfrastructureRef != nil {
+		infra, err = r.getCurrentInfrastructureClusterState(ctx, cluster)
+		if err != nil {
+			return nil, errors.Wrap(err, "current cluster state not found")
+		}
+	}
+	// Reference to the control plane can be nil, and if the reference to the control plane is not nil the reference
+	// to the control plane infrastructure may or may not exist. If the linked ClusterClass requires control plane
+	// infrastructure this will throw an unrecoverable error due to the configuration mismatch.
+	if cluster.Spec.ControlPlaneRef != nil {
+		cp, err = r.getControlPlaneState(ctx, cluster, class)
+		if err != nil {
+			return nil, errors.Wrap(err, "current cluster state not found")
+		}
+	}
+	// This reference can be empty and should not return an error if there are not Machine Deployments associated with
+	// the cluster at time of requesting state. If the machine deployments are missing information about their own
+	// infrastructure and bootstrap references this will throw an error.
```

Comment:
nit
```suggestion
	// A cluster might have zero or n machine deployments (lt is expected to be zero on the first reconcile loop).
```
let's move the details about the machine deployment references when defining getCurrentMachineDeploymentState